Source/Core/VideoCommon/GraphicsModSystem/Runtime/GraphicsModManager.cpp

```diff
@@ -0,0 +1,262 @@
+        m_actions.push_back(create_action(feature.m_action, feature.m_action_data, group, mod));
+        std::visit(overloaded{
+                       [&](const DrawTextureTarget& the_target) {
+                         m_draw_target_to_actions[the_target.m_texture_info_string].push_back(
+                             m_actions.back().get());
+                       },
+                       [&](const LoadTextureTarget& the_target) {
+                         m_load_target_to_actions[the_target.m_texture_info_string].push_back(
+                             m_actions.back().get());
+                       },
+                       [&](const EFBTarget& the_target) {
+                         EFBInfo info;
+                         info.m_height = the_target.m_height;
+                         info.m_width = the_target.m_width;
+                         info.m_texture_format = the_target.m_texture_format;
+                         m_efb_target_to_actions[info].push_back(m_actions.back().get());
+                       },
+                       [&](const ProjectionTarget& the_target) {
+                         if (the_target.m_texture_info_string)
+                         {
+                           const auto lookup =
+                               fmt::format("{}_{}", *the_target.m_texture_info_string,
+                                           static_cast<int>(the_target.m_projection_type));
+                           m_projection_texture_target_to_actions[lookup].push_back(
+                               m_actions.back().get());
+                         }
+                         else
+                         {
+                           m_projection_target_to_actions[the_target.m_projection_type].push_back(
+                               m_actions.back().get());
+                         }
+                       },
+                   },
+                   target);
+      };
+
+      // Prefer groups in the pack over groups from another pack
+      if (const auto local_it = group_to_targets.find(internal_group);
+          local_it != group_to_targets.end())
+      {
+        for (const GraphicsTargetConfig& target : local_it->second)
+        {
+          add_target(target, &m_groups[internal_group], &mod);
+        }
+      }
+      else if (const auto global_it = group_to_targets.find(feature.m_group);
+               global_it != group_to_targets.end())
+      {
+        for (const GraphicsTargetConfig& target : global_it->second)
+        {
+          add_target(target, &m_groups[feature.m_group], &mod);
+        }
+      }
+      else
+      {
+        WARN_LOG_FMT(VIDEO, "Specified graphics mod group '{}' was not found for mod '{}'",
+                     feature.m_group, mod.m_title);
+      }
+    }
+  }
+}
+
+void GraphicsModManager::EndOfFrame()
+{
+  for (auto&& action : m_actions)
+  {
+    static_cast<DecoratedAction*>(action.get())->OnFrameEnd();
```

Comment:
This seems kind of like a code smell. Maybe a way that doesn't require down/up casting instead would be nicer?